[WPE][CMake] Allow disabling the build of WTR without disabling the Cog build and also control the platform plugins of Cog

clopez opened the pull request:
#### 7aa121b35701dc54d093db118de218057a034f6e
<pre>
[WPE][CMake] Allow disabling the build of WTR without disabling the Cog build and also control the platform plugins of Cog
<a href="https://bugs.webkit.org/show_bug.cgi?id=249036">https://bugs.webkit.org/show_bug.cgi?id=249036</a>

Reviewed by Michael Catanzaro.

When building WPE for the rpi3 using the script build-webkit with a
yocto-based-cross-toolchain I found this issues:

 - for mesa-vc4 target: gtk4 is not available so the
   build fails when cog tries to enable the gtk4 plugin
 - for userland target: wpebackend-fdo is not available (only wpebackend-rdk)
   and wpebackend-fdo is needed to build WebKitTestRunner, the API tests runner
   and the MiniBrowser.

To fix this issues this patch allows to have more control over how Cog is built
and also to disable the build of WTR, MiniBrowser and APITestRunner meanwhile
keeping the build of Cog enabled.

After this patch this is how you would build for this targets:

rpi3-mesa-vc4:
  -DENABLE_COG=ON -DWPE_COG_PLATFORMS=drm,headless,wayland
  So gtk4 is disabled

rpi3-userland:
  -DENABLE_MINIBROWSER=OFF -DENABLE_API_TESTS=OFF -DENABLE_LAYOUT_TESTS=OFF -DENABLE_COG=ON -DWPE_COG_PLATFORMS=none
  So only Cog gets built and also it builds without platform plugins (which require wpebackend-fdo)
  The keyword &apos;none&apos; is used as a special key in CMake to later pass to the Cog meson build
  the parameter &quot;-Dplatforms=&quot; which is needed to set the list of platforms to empty instead
  of using the default defined on Cog at meson_options.txt

This patch also fixes the build on a cross-build environment by passing the path
of the cross-toolchain pkg_config to the Cog build.

And finally it also changes the GTK port to start using also the new introduced
CMake option ENABLE_LAYOUT_TESTS for consistency.

* Source/cmake/OptionsGTK.cmake
* Source/cmake/OptionsWPE.cmake:
* Source/cmake/WebKitFeatures.cmake:
* Tools/PlatformGTK.cmake:
* Tools/PlatformWPE.cmake:

Canonical link: <a href="https://commits.webkit.org/257827@main">https://commits.webkit.org/257827@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7166b0ad6231aa4d360da70570345c4584429b14

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99580 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8755 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32672 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108942 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169179 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103575 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9342 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86063 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92053 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106859 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105350 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7061 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90581 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34023 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88877 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21934 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/77295 "Found 2 new API test failures: /WebKitGTK/TestContextMenu:/webkit/WebKitWebPage/context-menu-node, /WebKitGTK/TestWebKitUserContentFilterStore:/webkit/WebKitUserContentFilterStore/filter-save-load (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/90229 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2594 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23449 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/86118 "Built successfully") | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2528 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45840 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28930 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8673 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42924 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/88989 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4388 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19900 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2765 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Tools/PlatformWPE.cmake:
```diff
@@ -1,24 +1,37 @@
-if (DEVELOPER_MODE AND ENABLE_COG)
+if (ENABLE_COG)
     include(ExternalProject)
+    if ("${WPE_COG_PLATFORMS}" STREQUAL "")
```
I think this should be exposed as a string-valued option. What you have here is basically a secret only for those who know to use it.

clopez replied:
Yes, it seems it will be more clean doing it that way
I can give it a try

mcatanzaro replied:
Although I see now there are other preexisting "options" that should match: `WPE_COG_REPO` and `WPE_COG_TAG`.

clopez replied:
As far as I can see all the CMake variables that we expose are all boolean. I can't see another example of a a CMake string variable exposed in WebKit. And it doesn't look very right to start doing that just for this.

Do you know about any example of a string variable exposed? The closest I could find is the CMake variable `USER_AGENT_BRANDING` but that one is also hidden to the build

mcatanzaro replied:
So it seems CMake options are always bool value. Ignore me, then....

mcatanzaro commented on Tools/PlatformWPE.cmake:
```diff
@@ -1,24 +1,37 @@
+    elseif ("${WPE_COG_PLATFORMS}" STREQUAL "none")
+        set(WPE_COG_PLATFORMS "")
+    endif ()
+    if (DEFINED ENV{PKG_CONFIG_PATH})
+        set(WPE_COG_PKG_CONFIG_PATH ${CMAKE_BINARY_DIR}:$ENV{PKG_CONFIG_PATH})
+    else ()
+        set(WPE_COG_PKG_CONFIG_PATH ${CMAKE_BINARY_DIR})
+    endif ()
```
Hm, `${CMAKE_BINARY_DIR}` is notably the _toplevel_ build directory. Surely cog.pc will not be installed there. Is that telling pkg-config to search the _entire_ build directory for pkg-config files? Maybe this could be more targeted?

clopez replied:
It is actually searching for `wpe-webkit-1.0.pc` there. The idea is that we don't install WPE after building it, but instead tell the cog build system to look for the wpe.pc file in the build directory

clopez committed:
[WPE][CMake] Allow disabling the build of WTR without disabling the Cog build and also control the platform plugins of Cog
https://bugs.webkit.org/show_bug.cgi?id=249036

Reviewed by Michael Catanzaro.

When building WPE for the rpi3 using the script build-webkit with a
yocto-based-cross-toolchain I found this issues:

 - for mesa-vc4 target: gtk4 is not available so the
   build fails when cog tries to enable the gtk4 plugin
 - for userland target: wpebackend-fdo is not available (only wpebackend-rdk)
   and wpebackend-fdo is needed to build WebKitTestRunner, the API tests runner
   and the MiniBrowser.

To fix this issues this patch allows to have more control over how Cog is built
and also to disable the build of WTR, MiniBrowser and APITestRunner meanwhile
keeping the build of Cog enabled.

After this patch this is how you would build for this targets:

rpi3-mesa-vc4:
  -DENABLE_COG=ON -DWPE_COG_PLATFORMS=drm,headless,wayland
  So gtk4 is disabled

rpi3-userland:
  -DENABLE_MINIBROWSER=OFF -DENABLE_API_TESTS=OFF -DENABLE_LAYOUT_TESTS=OFF -DENABLE_COG=ON -DWPE_COG_PLATFORMS=none
  So only Cog gets built and also it builds without platform plugins (which require wpebackend-fdo)
  The keyword 'none' is used as a special key in CMake to later pass to the Cog meson build
  the parameter "-Dplatforms=" which is needed to set the list of platforms to empty instead
  of using the default defined on Cog at meson_options.txt

This patch also fixes the build on a cross-build environment by passing the path
of the cross-toolchain pkg_config to the Cog build.

And finally it also changes the GTK port to start using also the new introduced
CMake option ENABLE_LAYOUT_TESTS for consistency.

* Source/cmake/OptionsGTK.cmake
* Source/cmake/OptionsWPE.cmake:
* Source/cmake/WebKitFeatures.cmake:
* Tools/PlatformGTK.cmake:
* Tools/PlatformWPE.cmake:

Canonical link: https://commits.webkit.org/257827@main
